src/sync/app.cpp

```diff
@@ -575,5 +580,35 @@ void App::login_with_credentials(const AppCredentials& credentials,
     }, handler);
 }
 
+void App::log_out(std::function<void (Optional<AppError>)> completion_block) const {
+    if (!current_user())
```

Comment:
There's still a race condition because `current_user()` can return different things (or nullptr) at each call in this method. I'd suggest storing the user it to a local variable and doing all checks against that one instance.

Reply:
In a later PR, I will likely just create a mutex in app that prevents parallelisation of these methods. But I will fix this for now.